app/controllers/measurements_controller.rb

```diff
@@ -2,47 +2,50 @@ class MeasurementsController < ApplicationController
   layout false
 
   def show
-    iteration_time = reference_time - params[:iteration].to_i * interval
-    cache_key = {
-      id: params[:id],
-      day: iteration_time.yday,
-      interval_number: interval_number(iteration_time)
-    }
-
-    @readings = Rails.cache.fetch(cache_key, expires_in: 25.hours) do
-      @measurement = Measurement.find(params[:id])
-      @measurement.
-        readings.
-        includes(:sensor).
-        where('time > ? AND time <= ?', from, to).
-        to_json(include: :sensor)
+    if params[:iteration].blank?
+      last_reading = Reading.where(measurement_id: params[:id]).last
+      readings = readings_json((last_reading&.time || Time.now),
+                               (2 * interval).minutes)
+    else
+      to = reference_time - (params[:iteration].to_i * interval).minutes
+      readings = readings_json(to, 25.hours)
     end
 
-    render json: @readings
+    render json: readings
   end
 
   private
 
+  def readings_json(to, expires_in)
+    cache_key = { id: params[:id], to: to&.to_f }
+
+    Rails.cache.fetch(cache_key, expires_in: expires_in) do
+      Reading.includes(:sensor).
+        where(measurement_id: params[:id]).
+        where('time > ? AND time <= ?', to - interval.minutes, to).
+        to_json(only: :value, include: { sensor: { only: [:lat, :long] } })
+    end
+  end
+
   def interval_number(time)
-    minutes = (time.seconds_since_midnight/60).to_i
+    minutes = (time.getutc.seconds_since_midnight/60).to_i
```

Comment:
This is not true :) Timezone is not stored in postgresql (run \d+ readings in psql console):

```
 time           | timestamp without time zone |                                                       | plain   |              | 
```

What is more rails is converting date into UTC while asking DB:

```
[8] pry(main)> t = Time.now
=> 2016-01-07 14:59:48 +0100
[9] pry(main)> Reading.where('time > ?', Time.now)
  Reading Load (0.4ms)  SELECT "readings".* FROM "readings" WHERE (time > '2016-01-07 13:59:54.609035')
```

So the usage of the uts here does not have any influence into query correctness and to_json results (utc is returned with and without this change :)). 

If we want to have timezone stored in DB we need to do something like this: https://github.com/ismop/dap/blob/master/db/migrate/20151020161114_add_time_zone_to_timestamp.rb. But I don't believe it is necessary. In DAP case we had a problem connected with 2 apps creating entries in DB one was storing time in timezone +1 (and timezone was truncated)  and other was assuming that utc was used. This is not a case here.